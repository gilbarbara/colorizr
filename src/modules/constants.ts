import { ColorKeysTuple, ColorModelKey } from '~/types';

export const COLOR_KEYS: Record<ColorModelKey, ColorKeysTuple> = {
  hsl: ['h', 's', 'l'],
  oklab: ['l', 'a', 'b'],
  oklch: ['l', 'c', 'h'],
  rgb: ['r', 'g', 'b'],
};
export const COLOR_MODELS: ColorModelKey[] = ['hsl', 'oklab', 'oklch', 'rgb'];
export const DEG2RAD = Math.PI / 180;
export const LAB_TO_LMS = {
  l: [0.3963377773761749, 0.2158037573099136],
  m: [-0.1055613458156586, -0.0638541728258133],
  s: [-0.0894841775298119, -1.2914855480194092],
};
export const LRGB_TO_LMS = {
  l: [0.4122214708, 0.5363325363, 0.0514459929],
  m: [0.2119034982, 0.6806995451, 0.1073969566],
  s: [0.0883024619, 0.2817188376, 0.6299787005],
};
export const LSM_TO_LAB = {
  l: [0.2104542553, 0.793617785, 0.0040720468],
  a: [1.9779984951, 2.428592205, 0.4505937099],
  b: [0.0259040371, 0.7827717662, 0.808675766],
};
export const LSM_TO_RGB = {
  r: [4.0767416360759583, -3.3077115392580629, 0.2309699031821043],
  g: [-1.2684379732850315, 2.6097573492876882, -0.341319376002657],
  b: [-0.0041960761386756, -0.7034186179359362, 1.7076146940746117],
};
export const PRECISION = 5;
export const RAD2DEG = 180 / Math.PI;

export const MESSAGES = {
  alpha: 'amount must be a number between 0 and 1',
  left: 'left is required and must be a string',
  right: 'right is required and must be a string',
  input: 'input is required',
  inputHex: 'input is required and must be a hex',
  inputNumber: 'input is required and must be a number',
  inputString: 'input is required and must be a string',
  invalid: 'invalid input',
  invalidCSS: 'invalid CSS string',
  options: 'invalid options',
  threshold: 'threshold must be a number between 0 and 255',
} as const;
