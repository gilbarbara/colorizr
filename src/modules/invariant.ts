export function invariant(condition: boolean, message: string): asserts condition {
  if (condition) {
    return;
  }

  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production') {
    if (message === undefined) {
      throw new Error('invariant requires an error message argument');
    }
  }

  const error = !message
    ? new Error(
        'Minified exception occurred; use the non-minified dev environment for the full error message and additional helpful warnings.',
      )
    : new Error(message);

  error.name = 'colorizr';

  throw error;
}
