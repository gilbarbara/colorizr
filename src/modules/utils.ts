import { HSL, PlainObject, RGB, RGBArray } from '../types';

export const HSLKeys = ['h', 's', 'l'];
export const RGBKeys = ['r', 'g', 'b'];

/**
 * Constrain value into the range
 */
export function constrain(input: number, amount: number, range: number[], sign: string): number {
  invariant(arguments.length === 4, 'All parameters are required');

  const [min, max] = range;
  let value = expr(input + sign + amount);

  if (value < min) {
    value = min;
  } else if (value > max) {
    value = max;
  }

  return Math.abs(value);
}

/**
 * Constrain an angle
 */
export function constrainDegrees(input: number, amount: number): number {
  invariant(isNumber(input), 'input is required');

  let value = input + amount;

  if (value > 360) {
    value %= 360;
  }

  if (value < 0) {
    value += 360;
  }

  return Math.abs(value);
}

/**
 * Parse math string expressions
 */
export function expr(input: string): number {
  const chars = [...input];
  const n: string[] = [];
  const op: string[] = [];

  let parsed;
  let index = 0;
  let last = true;

  n[index] = '';

  // Parse the string
  for (const char of chars) {
    if (Number.isNaN(parseInt(char, 10)) && char !== '.' && !last) {
      op[index] = char;
      index++;
      n[index] = '';
      last = true;
    } else {
      n[index] += char;
      last = false;
    }
  }

  // Calculate the expression
  parsed = parseFloat(n[0]);

  for (const [o, element] of op.entries()) {
    const value = parseFloat(n[o + 1]);

    switch (element) {
      case '+':
        parsed += value;
        break;
      case '-':
        parsed -= value;
        break;
      case '*':
        parsed *= value;
        break;
      case '/':
        parsed /= value;
        break;
      default:
        break;
    }
  }

  return parsed;
}

export function invariant(condition: boolean, message: string): asserts condition {
  if (condition) {
    return;
  }

  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production') {
    if (message === undefined) {
      throw new Error('invariant requires an error message argument');
    }
  }

  let error;

  if (!message) {
    throw new Error(
      'Minified exception occurred; use the non-minified dev environment ' +
        'for the full error message and additional helpful warnings.',
    );
  } else {
    error = new Error(message);
  }

  error.name = 'colorizr';

  throw error;
}

/**
 * Check if an object contains HSL values
 */
export function isHSL(input: any): input is HSL {
  if (!isPlainObject(input)) {
    return false;
  }

  const entries = Object.entries(input);

  return (
    !!entries.length &&
    entries.every(
      ([key, value]) => HSLKeys.includes(key) && value >= 0 && value <= (key === 'h' ? 360 : 100),
    )
  );
}

/**
 * Check if the input is a number and not NaN
 */
export function isNumber(input: any): input is number {
  return typeof input === 'number' && !Number.isNaN(input);
}

/**
 * Check if the input is an object
 */
export function isPlainObject(input: any): input is PlainObject {
  if (!input) {
    return false;
  }

  const { toString } = Object.prototype;
  const prototype = Object.getPrototypeOf(input);

  return (
    toString.call(input) === '[object Object]' &&
    (prototype === null || prototype === Object.getPrototypeOf({}))
  );
}

/**
 * Check if an object contains RGB values.
 */
export function isRGB(input: any): input is RGB {
  if (!isPlainObject(input)) {
    return false;
  }

  const entries = Object.entries(input);

  return (
    !!entries.length &&
    entries.every(([key, value]) => RGBKeys.includes(key) && value >= 0 && value <= 255)
  );
}

/**
 * Check if an array contains RGB values.
 */
export function isRGBArray(input: any): input is RGBArray {
  return Array.isArray(input) && input.length === 3 && input.every(d => d >= 0 && d <= 255);
}

/**
 * Check if the input is a string
 */
export function isString(input: any): input is string {
  return typeof input === 'string';
}

/**
 * Limit values per type.
 */
export function limit(input: number, type: string): number {
  invariant(isNumber(input), 'Input is not a number');

  /* istanbul ignore else */
  if (RGBKeys.includes(type)) {
    return Math.max(Math.min(input, 255), 0);
  }

  if (['s', 'l'].includes(type)) {
    return Math.max(Math.min(input, 100), 0);
  }

  if (type === 'h') {
    return Math.max(Math.min(input, 360), 0);
  }

  throw new Error('Invalid type');
}

export const messages = {
  amount: 'amount must be a number',
  left: 'left is required and must be a string',
  right: 'right is required and must be a string',
  input: 'input is required',
  inputString: 'input is required and must be a string',
  invalid: 'invalid input',
  options: 'invalid options',
};

/**
 * Creates an object composed of the picked source properties.
 */
export function pick(input: PlainObject, options: string[]): PlainObject {
  if (!Array.isArray(options)) {
    throw new TypeError('options must be an array');
  }

  return options
    .filter(d => typeof input[d] !== 'undefined')
    .reduce((acc: PlainObject, d) => {
      acc[d] = input[d];

      return acc;
    }, {});
}

/**
 * Round decimal numbers.
 */
export function round(input: number, digits = 2): number {
  const factor = 10 ** digits;

  return Math.round(input * factor) / factor;
}
