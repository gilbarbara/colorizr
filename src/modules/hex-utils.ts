import { MESSAGES } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import { round } from '~/modules/utils';
import { isHex, isNumber } from '~/modules/validators';

import { Alpha } from '~/types';

/**
 * Add an alpha value to a hex string
 */
export function addAlphaToHex(input: string, alpha: Alpha): string {
  invariant(isHex(input), MESSAGES.inputHex);
  invariant(isNumber(alpha), MESSAGES.inputNumber);

  if (alpha >= 1) {
    return removeAlphaFromHex(input);
  }

  return `${removeAlphaFromHex(input)}${convertAlphaToHex(alpha)}`;
}

/**
 * Convert an alpha value to a hex value.
 */
export function convertAlphaToHex(input: Alpha): string {
  invariant(isNumber(input), MESSAGES.inputNumber);

  let alpha = input;

  if (input > 1) {
    alpha /= 100;
  }

  return Math.round(alpha * 255)
    .toString(16)
    .padStart(2, '0');
}

/**
 * Extract the alpha value from a hex string
 */
export function extractAlphaFromHex(input: string): number {
  invariant(isHex(input), MESSAGES.inputString);

  const alpha = input.substring(7, 9);

  if (!alpha) {
    return 1;
  }

  return round(parseInt(alpha, 16) / 255);
}

export function hexadecimalToNumber(input: string) {
  return round(parseInt(input, 16));
}

/**
 * Remove the alpha value from a hex string
 */
export function removeAlphaFromHex(input: string) {
  invariant(isHex(input), MESSAGES.inputHex);

  if (input.length === 5) {
    return input.substring(0, 4);
  }

  return input.substring(0, 7);
}
