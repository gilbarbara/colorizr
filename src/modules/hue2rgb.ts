import { invariant, isNumber, round } from './utils';

/**
 * Convert hue to RGB using chroma and median point
 */
export default function hue2rgb(point: number, chroma: number, h: number): number {
  invariant(isNumber(point) && isNumber(chroma) && isNumber(h), 'point, chroma and h are required');
  let hue = h;

  if (hue < 0) {
    hue += 1;
  }

  if (hue > 1) {
    hue -= 1;
  }

  if (hue < 1 / 6) {
    return round(point + (chroma - point) * 6 * hue, 4);
  }

  if (hue < 1 / 2) {
    return round(chroma, 4);
  }

  if (hue < 2 / 3) {
    return round(point + (chroma - point) * (2 / 3 - hue) * 6, 4);
  }

  return round(point, 4);
}
