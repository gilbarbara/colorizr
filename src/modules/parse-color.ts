import { invariant, isHSL, isPlainObject, isRGB, isString, limit, messages } from './utils';

import hex2hsl from '../hex2hsl';
import hex2rgb from '../hex2rgb';
import hsl2hex from '../hsl2hex';
import hsl2rgb from '../hsl2rgb';
import isValidHex from '../is-valid-hex';
import parseCSS from '../parse-css';
import rgb2hex from '../rgb2hex';
import rgb2hsl from '../rgb2hsl';
import { Colors, HSL, PlainObject, RGB, RGBArray } from '../types';

export default function parseColor(color: string | HSL | RGB | RGBArray): Colors {
  invariant(!!color, messages.input);

  const output: PlainObject = {};

  if (isString(color)) {
    const hex = parseCSS(color) as string;

    invariant(isValidHex(hex), 'input is not valid');

    output.hex = hex;
    output.rgb = hex2rgb(hex);
    output.hsl = hex2hsl(hex);
  } else if (Array.isArray(color)) {
    output.rgb = {
      r: limit(color[0], 'r'),
      g: limit(color[1], 'g'),
      b: limit(color[2], 'b'),
    };

    output.hex = rgb2hex(output.rgb);
    output.hsl = rgb2hsl(output.rgb);
  } else if (isPlainObject(color)) {
    if (isHSL(color)) {
      output.hsl = {
        h: limit(color.h, 'h'),
        s: limit(color.s, 's'),
        l: limit(color.l, 'l'),
      };
      output.rgb = hsl2rgb(output.hsl);
    } else if (isRGB(color)) {
      output.rgb = {
        r: limit(color.r, 'r'),
        g: limit(color.g, 'g'),
        b: limit(color.b, 'b'),
      };
      output.hsl = rgb2hsl(output.rgb);
    } else {
      throw new Error('invalid color');
    }

    output.hex = hsl2hex(output.hsl);
  } else {
    throw new Error(messages.input);
  }

  return output as Colors;
}
