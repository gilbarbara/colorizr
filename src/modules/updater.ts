import { constrain, invariant, isNumber, isString, messages } from './utils';

import hex2hsl from '../hex2hsl';
import parseCSS from '../parse-css';
import { shift } from '../shift';

/**
 * Update color properties
 */
export default function updater(type: 'h' | 's' | 'l', sign: '+' | '-') {
  return (input: string, amount: number) => {
    invariant(isString(input), messages.inputString);
    invariant(isNumber(amount), messages.amount);

    const hex = parseCSS(input);
    const hsl = hex2hsl(hex);

    return shift(hex, {
      [type]: constrain(hsl[type], amount, [0, 100], sign),
    });
  };
}
