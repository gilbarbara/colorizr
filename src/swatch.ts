import { MESSAGES } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import { clamp } from '~/modules/utils';
import { isHex, isNamedColor, isString } from '~/modules/validators';

import convert from '~/convert';
import { oklch2hex } from '~/converters';
import oklch2rgb from '~/converters/oklch2rgb';
import rgb2hex from '~/converters/rgb2hex';
import extractColorParts from '~/extract-color-parts';
import parseCSS from '~/parse-css';
import { ColorTokens, ColorType, HEX, LCH } from '~/types';

export type Swatch = {
  [key in ColorTokens]: string;
};

export type SwatchVariant = 'deep' | 'neutral' | 'pastel' | 'subtle' | 'vibrant';

export interface SwatchOptions {
  /**
   * Output color format.
   *
   * If not specified, the output will use the same format as the input color.
   */
  format?: ColorType;
  /**
   * The mode of the swatch.
   * - 'light': Focuses on lighter tones starting from a higher lightness.
   * - 'dark': Focuses on darker tones starting from a lower lightness.
   *
   * If omitted, a balanced palette is generated.
   */
  mode?: 'light' | 'dark';
  /**
   * Generate a monochromatic swatch.
   *
   * Disable chroma adjustments for all shades.
   * @default false
   */
  monochromatic?: boolean;
  /**
   * Determines the scale type for the swatch.
   * - 'linear': Shades are distributed with equal lightness intervals.
   * - 'dynamic': Shades are distributed adaptively based on the input color.
   * @default 'dynamic'
   */
  scale?: 'dynamic' | 'linear';
  /**
   * The variant of the swatch.
   * - 'deep': Generates rich and bold tones with significantly reduced lightness.
   * - 'neutral': Generates muted tones by reducing chroma.
   * - 'pastel': Produces soft and airy tones with significant chroma reduction.
   * - 'subtle': Creates extremely desaturated tones, close to grayscale.
   * - 'vibrant': Enhances chroma for bold and striking tones.
   */
  variant?: SwatchVariant;
}

const MIN_LIGHTNESS = 5;
const MAX_LIGHTNESS = 95;

/**
 * Generate a shade of a color based its lightness tuning factor
 */
function shadeColorDynamic(input: LCH, lightnessTuningFactor: number, chromaFactor = 0): HEX {
  if (lightnessTuningFactor === 0) {
    const { l } = input;

    return rgb2hex(oklch2rgb({ ...input, l: l / 100 }));
  }

  // Convert back to RGB and make sure it's within the sRGB gamut
  return shadeColor(input, input.l + lightnessTuningFactor, chromaFactor);
}

/**
 * Generate a shade of a color based its lightness tuning factor
 */
function shadeColor(input: LCH, lightness: number, chromaFactor = 0): HEX {
  const { c, h } = input;

  const adjustedChroma = clamp(c + chromaFactor, 0, 0.4);

  // Convert back to RGB and make sure it's within the sRGB gamut
  return oklch2hex({ l: lightness / 100, c: adjustedChroma, h });
}

/**
 * Generate a palette of shades of a color
 */
export default function swatch(input: string, options: SwatchOptions = {}): Swatch {
  invariant(isString(input), MESSAGES.inputString);
  const { format, monochromatic = false, scale = 'dynamic', mode, variant = 'base' } = options;

  const lch = parseCSS(input, 'oklch');
  const chromaScale = {
    base: 1,
    deep: 0.8,
    neutral: 0.5,
    pastel: 0.3,
    subtle: 0.2,
    vibrant: 1.25,
  }[variant];

  lch.l = 50;
  lch.c *= chromaScale;

  if (variant === 'deep') {
    lch.l *= 0.7;
  }

  const colorFormat = isHex(input) || isNamedColor(input) ? 'hex' : extractColorParts(input).model;

  const currentLightness = lch.l;
  let lightSteps = 5;
  let darkSteps = 8;

  if (mode) {
    lightSteps *= mode === 'light' ? 0.75 : 1.25;
    darkSteps *= mode === 'light' ? 1.25 : 0.75;
  }

  const lightStepSize = (MAX_LIGHTNESS - currentLightness) / lightSteps;
  const darkStepSize = (-1 * (currentLightness - MIN_LIGHTNESS)) / darkSteps;

  const getChromaFactor = (value: number): number => {
    return monochromatic ? 0 : value; // Disable adjustments for monochromatic swatches
  };

  const output: Swatch =
    scale === 'linear'
      ? {
          50: shadeColor(lch, 95, getChromaFactor(-0.00375)),
          100: shadeColor(lch, 90, getChromaFactor(-0.00375)),
          200: shadeColor(lch, 80, getChromaFactor(-0.00375)),
          300: shadeColor(lch, 70, getChromaFactor(-0.00375)),
          400: shadeColor(lch, 60, getChromaFactor(-0.00375)),
          500: shadeColor(lch, 50),
          600: shadeColor(lch, 40, getChromaFactor(0.025)),
          700: shadeColor(lch, 30, getChromaFactor(0.05)),
          800: shadeColor(lch, 20, getChromaFactor(0.075)),
          900: shadeColor(lch, 10, getChromaFactor(0.1)),
        }
      : {
          50: shadeColorDynamic(lch, 5 * lightStepSize, getChromaFactor(-0.00375)),
          100: shadeColorDynamic(lch, 4 * lightStepSize, getChromaFactor(-0.00375)),
          200: shadeColorDynamic(lch, 3 * lightStepSize, getChromaFactor(-0.00375)),
          300: shadeColorDynamic(lch, 2 * lightStepSize, getChromaFactor(-0.00375)),
          400: shadeColorDynamic(lch, lightStepSize, getChromaFactor(-0.00375)),
          500: shadeColorDynamic(lch, 0),
          600: shadeColorDynamic(lch, 1.6 * darkStepSize, getChromaFactor(0.025)),
          700: shadeColorDynamic(lch, 3.2 * darkStepSize, getChromaFactor(0.05)),
          800: shadeColorDynamic(lch, 4.8 * darkStepSize, getChromaFactor(0.075)),
          900: shadeColorDynamic(lch, 6.4 * darkStepSize, getChromaFactor(0.1)),
        };

  return Object.entries(output).reduce((acc, [key, value]) => {
    return {
      ...acc,
      [key]: convert(value, format ?? colorFormat),
    };
  }, {} as Swatch);
}
