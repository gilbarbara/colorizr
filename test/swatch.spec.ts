import { MESSAGES } from '~/modules/constants';

import swatch, { SwatchOptions } from '~/swatch';

import { brightPink, green, orange, violet, yellow } from './__fixtures__';

describe('swatch', () => {
  it.each([
    [brightPink.hex, undefined],
    [brightPink.hex, { mode: 'light' }],
    [brightPink.hex, { mode: 'dark' }],
    [brightPink.hex, { format: 'oklch' }],
    [green.hslString, undefined],
    [green.hslString, { scale: 'linear' }],
    [green.hslString, { variant: 'deep' }],
    [orange.oklabString, undefined],
    [orange.oklabString, { variant: 'neutral' }],
    [violet.oklchString, undefined],
    [violet.oklchString, { variant: 'pastel' }],
    [yellow.rgbString, undefined],
    [yellow.rgbString, { variant: 'subtle' }],
    ['oklch(0.65 0.3 27.34)', undefined],
    ['#d4ffc7', { variant: 'vibrant' }],
    ['#808080', { monochromatic: true }],
    ['#808080', { monochromatic: true, scale: 'linear' }],
  ] as Array<[string, SwatchOptions]>)(
    'should return properly with %s and %s',
    (input, variant) => {
      expect(swatch(input, variant)).toMatchSnapshot();
    },
  );

  it('should fail with invalid parameters', () => {
    // @ts-expect-error - invalid parameters
    expect(() => swatch(undefined)).toThrow(MESSAGES.inputString);
  });
});
