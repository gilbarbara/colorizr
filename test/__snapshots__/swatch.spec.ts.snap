// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`swatch > should return properly with #808080 and { monochromatic: true } 1`] = `
{
  "100": "#d1d1d1",
  "200": "#b4b4b4",
  "300": "#989898",
  "400": "#7d7d7d",
  "50": "#eeeeee",
  "500": "#636363",
  "600": "#4a4a4a",
  "700": "#333333",
  "800": "#1d1d1d",
  "900": "#090909",
}
`;

exports[`swatch > should return properly with #808080 and { monochromatic: true, scale: 'linear' } 1`] = `
{
  "100": "#dedede",
  "200": "#bebebe",
  "300": "#9e9e9e",
  "400": "#808080",
  "50": "#eeeeee",
  "500": "#636363",
  "600": "#484848",
  "700": "#2e2e2e",
  "800": "#161616",
  "900": "#030303",
}
`;

exports[`swatch > should return properly with #d4ffc7 and { variant: 'vibrant' } 1`] = `
{
  "100": "#afe29f",
  "200": "#93c483",
  "300": "#77a868",
  "400": "#5d8c4e",
  "50": "#ccffbc",
  "500": "#427132",
  "600": "#1c5a00",
  "700": "#004200",
  "800": "#002b00",
  "900": "#001400",
}
`;

exports[`swatch > should return properly with #ff0044 and { format: 'oklch' } 1`] = `
{
  "100": "oklch(73.814% 0.16213 14.45451)",
  "200": "oklch(69.005% 0.20179 16.21216)",
  "300": "oklch(65.108% 0.2365 18.79042)",
  "400": "oklch(59.786% 0.24055 21.28701)",
  "50": "oklch(79.349% 0.12138 11.90057)",
  "500": "oklch(53.396% 0.21706 26.08696)",
  "600": "oklch(48.149% 0.19757 29.23375)",
  "700": "oklch(42.475% 0.1743 29.23276)",
  "800": "oklch(36.391% 0.14933 29.23423)",
  "900": "oklch(29.589% 0.12142 29.23565)",
}
`;

exports[`swatch > should return properly with #ff0044 and { mode: 'dark' } 1`] = `
{
  "100": "#ff5f74",
  "200": "#ff425e",
  "300": "#ff1949",
  "400": "#e70033",
  "50": "#ff7a8a",
  "500": "#cd001b",
  "600": "#a80000",
  "700": "#800000",
  "800": "#560000",
  "900": "#290000",
}
`;

exports[`swatch > should return properly with #ff0044 and { mode: 'light' } 1`] = `
{
  "100": "#ffa5b0",
  "200": "#ff7a8a",
  "300": "#ff4c65",
  "400": "#f90041",
  "50": "#ffd0d8",
  "500": "#cd001b",
  "600": "#ba0000",
  "700": "#a50000",
  "800": "#900000",
  "900": "#790000",
}
`;

exports[`swatch > should return properly with #ff0044 and undefined 1`] = `
{
  "100": "#ff7a8a",
  "200": "#ff586f",
  "300": "#ff3053",
  "400": "#ed0039",
  "50": "#ff9aa7",
  "500": "#cd001b",
  "600": "#b30000",
  "700": "#970000",
  "800": "#7a0000",
  "900": "#5b0000",
}
`;

exports[`swatch > should return properly with hsl(136 100% 50%) and { scale: 'linear' } 1`] = `
{
  "100": "hsl(129.71 100% 60%)",
  "200": "hsl(129.87 100% 45.29%)",
  "300": "hsl(120 100% 38.63%)",
  "400": "hsl(120 100% 32.16%)",
  "50": "hsl(128.18 100% 65.49%)",
  "500": "hsl(120 100% 25.88%)",
  "600": "hsl(120 100% 20%)",
  "700": "hsl(120 100% 14.12%)",
  "800": "hsl(120 100% 8.43%)",
  "900": "hsl(120 100% 3.33%)",
}
`;

exports[`swatch > should return properly with hsl(136 100% 50%) and { variant: 'deep' } 1`] = `
{
  "100": "hsl(124.9 77.78% 62.94%)",
  "200": "hsl(130.18 73.99% 43.73%)",
  "300": "hsl(120 100% 30.39%)",
  "400": "hsl(120 100% 23.14%)",
  "50": "hsl(123.87 100% 75.69%)",
  "500": "hsl(120 100% 16.08%)",
  "600": "hsl(120 100% 12.75%)",
  "700": "hsl(120 100% 9.61%)",
  "800": "hsl(120 100% 6.27%)",
  "900": "hsl(120 100% 3.33%)",
}
`;

exports[`swatch > should return properly with hsl(136 100% 50%) and undefined 1`] = `
{
  "100": "hsl(135 96.83% 50.59%)",
  "200": "hsl(124.07 100% 43.33%)",
  "300": "hsl(120 100% 37.25%)",
  "400": "hsl(120 100% 31.57%)",
  "50": "hsl(128.18 100% 65.49%)",
  "500": "hsl(120 100% 25.88%)",
  "600": "hsl(120 100% 20.59%)",
  "700": "hsl(120 100% 15.49%)",
  "800": "hsl(120 100% 10.2%)",
  "900": "hsl(120 100% 5.1%)",
}
`;

exports[`swatch > should return properly with oklab(70.622% 0.1374 0.14283) and { variant: 'neutral' } 1`] = `
{
  "100": "oklab(85.291% 0.05809 0.06621)",
  "200": "oklab(76.912% 0.06605 0.06864)",
  "300": "oklab(67.889% 0.06645 0.06852)",
  "400": "oklab(58.943% 0.06709 0.0684)",
  "50": "oklab(91.316% 0.02509 0.05462)",
  "500": "oklab(50% 0.06984 0.07108)",
  "600": "oklab(41.215% 0.08844 0.08327)",
  "700": "oklab(32.704% 0.11711 0.06554)",
  "800": "oklab(27.542% 0.09863 0.0552)",
  "900": "oklab(21.638% 0.07748 0.04336)",
}
`;

exports[`swatch > should return properly with oklab(70.622% 0.1374 0.14283) and undefined 1`] = `
{
  "100": "oklab(79.198% 0.07597 0.12194)",
  "200": "oklab(73.982% 0.11238 0.13328)",
  "300": "oklab(68.043% 0.13555 0.13758)",
  "400": "oklab(59.235% 0.14269 0.1195)",
  "50": "oklab(84.977% 0.03983 0.1112)",
  "500": "oklab(50.44% 0.15441 0.10138)",
  "600": "oklab(43.91% 0.15724 0.088)",
  "700": "oklab(38.726% 0.13867 0.07761)",
  "800": "oklab(33.143% 0.11868 0.06642)",
  "900": "oklab(26.389% 0.0945 0.05289)",
}
`;

exports[`swatch > should return properly with oklch(0.65 0.3 27.34) and undefined 1`] = `
{
  "100": "oklch(69.36% 0.19638 26.36542)",
  "200": "oklch(64.788% 0.23788 26.12125)",
  "300": "oklch(62.796% 0.25768 29.23494)",
  "400": "oklch(62.796% 0.25768 29.23494)",
  "50": "oklch(74.836% 0.15206 27.95369)",
  "500": "oklch(56.362% 0.23129 29.23266)",
  "600": "oklch(51.12% 0.20977 29.235)",
  "700": "oklch(45.334% 0.18603 29.2326)",
  "800": "oklch(38.937% 0.15978 29.23293)",
  "900": "oklch(31.822% 0.13058 29.23275)",
}
`;

exports[`swatch > should return properly with oklch(47.642% 0.29956 274.93693) and { variant: 'pastel' } 1`] = `
{
  "100": "oklch(85.528% 0.07122 273.43732)",
  "200": "oklch(76.966% 0.08707 275.4238)",
  "300": "oklch(67.926% 0.08696 274.86158)",
  "400": "oklch(59.004% 0.08682 274.92258)",
  "50": "oklch(93.566% 0.03103 257.22684)",
  "500": "oklch(49.871% 0.09017 274.96258)",
  "600": "oklch(40.999% 0.11411 275.20898)",
  "700": "oklch(32.002% 0.14028 275.03025)",
  "800": "oklch(24.427% 0.1541 274.52589)",
  "900": "oklch(21.803% 0.13504 276.86451)",
}
`;

exports[`swatch > should return properly with oklch(47.642% 0.29956 274.93693) and undefined 1`] = `
{
  "100": "oklch(77.724% 0.11327 274.07048)",
  "200": "oklch(69.778% 0.16006 277.47755)",
  "300": "oklch(62.286% 0.20608 277.97715)",
  "400": "oklch(55.313% 0.25032 277.38138)",
  "50": "oklch(85.977% 0.0686 264.91522)",
  "500": "oklch(48.983% 0.29101 275.61636)",
  "600": "oklch(45.903% 0.29037 274.18684)",
  "700": "oklch(43.228% 0.27151 275.07769)",
  "800": "oklch(40.859% 0.24999 278.47206)",
  "900": "oklch(38.483% 0.22875 282.47665)",
}
`;

exports[`swatch > should return properly with rgb(255 230 109) and { variant: 'subtle' } 1`] = `
{
  "100": "rgb(213 209 191)",
  "200": "rgb(184 181 163)",
  "300": "rgb(156 153 136)",
  "400": "rgb(129 126 109)",
  "50": "rgb(243 239 220)",
  "500": "rgb(103 100 81)",
  "600": "rgb(82 75 39)",
  "700": "rgb(62 50 0)",
  "800": "rgb(43 26 0)",
  "900": "rgb(24 3 0)",
}
`;

exports[`swatch > should return properly with rgb(255 230 109) and undefined 1`] = `
{
  "100": "rgb(234 209 92)",
  "200": "rgb(204 180 58)",
  "300": "rgb(176 152 8)",
  "400": "rgb(148 124 0)",
  "50": "rgb(255 239 123)",
  "500": "rgb(122 97 0)",
  "600": "rgb(100 70 0)",
  "700": "rgb(78 42 0)",
  "800": "rgb(57 13 0)",
  "900": "rgb(33 0 0)",
}
`;
